Implement deferred interaction triggers

crisbeto opened the pull request:
Includes a few prerequisites, as well as the actual implementation, for the `on interaction` and `prefetch on interaction` triggers. The changes here will be the base for implementing the `on viewport` and `on hover` triggers as well. Includes the following commits:

### refactor(compiler): extract deferred block trigger information
Reworks the compiler to use the API introduced in https://github.com/angular/angular/pull/51816 to match triggers to the element nodes they point to. This will be used to generate the new instructions for `on interaction` and `prefetch on interaction`.

### refactor(compiler): implement final instruction generation for interaction triggers
Updates the logic that generates the instructions for the `on interaction` and `prefetch on interaction` triggers to their final shape. Now the instructions take two arguments:
1. `triggerIndex` - index at which to find the trigger in the view where it will be rendered.
2. `walkUpTimes` - tells the runtime how many views up it needs to go to find the trigger element. If the argument is omitted, it means that the trigger is in the same view as the deferred block. A positive number means that the runtime needs to go up X amount of times to find the trigger. A negative number means that the trigger is inside the root view of the placeholder block. Negative numbers are capped at -1 since the placeholder is always in the same position at runtime.

### feat(core): implement deferred block interaction triggers
Adds the implementation for the `on interaction` and `prefetch on interaction` triggers.

crisbeto committed:
refactor(compiler): extract deferred block trigger information

Reworks the compiler to use the API introduced in #51816 to match triggers to the element nodes they point to. This will be used to generate the new instructions for `on interaction` and `prefetch on interaction`.

crisbeto committed:
refactor(compiler): implement final instruction generation for interaction triggers

Updates the logic that generates the instructions for the `on interaction` and `prefetch on interaction` triggers to their final shape. Now the instructions take two arguments:
1. `triggerIndex` - index at which to find the trigger in the view where it will be rendered.
2. `walkUpTimes` - tells the runtime how many views up it needs to go to find the trigger element. If the argument is omitted, it means that the trigger is in the same view as the deferred block. A positive number means that the runtime needs to go up X amount of times to find the trigger. A negative number means that the trigger is inside the root view of the placeholder block. Negative numbers are capped at -1 since the placeholder is always in the same position at runtime.

crisbeto commented on packages/core/src/render3/instructions/defer_event_manager.ts:
```diff
@@ -0,0 +1,79 @@
+/** Manages events related to deferred blocks. */
+export class ɵDeferEventManager {
+  /** @nocollapse */
+  static ɵprov = /** @pureOrBreakMyCode */ ɵɵdefineInjectable({
```
I made this class an injectable, because we'll need access to `NgZone` for some of the other triggers.

crisbeto commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -33,7 +35,7 @@ import {ɵɵtemplate} from './template';
  */
 function shouldTriggerDeferBlock(injector: Injector): boolean {
-  const config = injector.get(DEFER_BLOCK_CONFIG, {optional: true});
+  const config = injector.get(DEFER_BLOCK_CONFIG, null, {optional: true});
```
Unrelated change, I just noticed that we were using a deprecated signature.

crisbeto commented on packages/core/test/acceptance/defer_spec.ts:
```diff
@@ -1009,4 +1009,471 @@ describe('#defer', () => {
+         expect(fixture.nativeElement.textContent.trim()).toBe('Main content');
+       }));
+
+    it('should resolve a trigger on a component outside the defer block', fakeAsync(() => {
```
I had to use `fakeAsync` here, because I couldn't get the `afterRender` to flush reliably in the browser tests.

crisbeto commented on packages/core/test/acceptance/defer_spec.ts:
```diff
@@ -1009,4 +1009,471 @@ describe('#defer', () => {
+  });
+
+  describe('interaction triggers', () => {
+    it('should load the deferred content when the trigger is clicked', fakeAsync(() => {
```
I had to use `fakeAsync` here, because I couldn't get the `afterRender` to fire predictably in the browser tests.

AndrewKushnir replied:
Would it be possible to repro this in `main` using `afterRender` and a similar component structure? It'd be interesting to investigate to make sure there are no issues with `afterRender` timing.

crisbeto replied:
From my investigation, `afterRender` was definitely firing. The problem was that it's asynchronous and we need it to fire before the relevant assertion. I'm not sure that the `defer` tests should be concerned with the `afterRender` timing. That's why `afterRender` has tests of its own.

AndrewKushnir commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
+    const renderedState = lDetails[DEFER_BLOCK_STATE];
+
+    // If the block was loaded before the trigger was resolved, we don't need to do anything.
+    if (renderedState !== DeferBlockInternalState.Initial &&
+        renderedState !== DeferBlockState.Placeholder) {
+      afterRenderRef.destroy();
+      return;
+    }
```
I'd propose moving this check up and call `getTriggerLView` after this check, so we don't lookup trigger LView in wrong (non-placeholder) views.

AndrewKushnir commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
+  // If the value is negative, it means that the trigger is inside the placeholder.
+  const deferredContainer = deferredHostLView[deferredTNode.index];
+  ngDevMode && assertLContainer(deferredContainer);
+  const triggerLView = deferredContainer[CONTAINER_HEADER_OFFSET] ?? null;
+
```
I'd propose to also check that the LView in this container is indeed a placeholder one.

```
    if (ngDevMode) {
      const lDetails = getLDeferBlockDetails(deferredHostLView, deferredTNode);
      const renderedState = lDetails[DEFER_BLOCK_STATE];
      assertEqual(renderedState, DeferBlockState.Placeholder,
        'Expected a placeholder to be rendered in this defer block.');
    }
```

AndrewKushnir commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
+  const triggerLView = deferredContainer[CONTAINER_HEADER_OFFSET] ?? null;
+
+  // We need to null check, because the placeholder might not have been rendered yet.
+  if (ngDevMode && triggerLView != null) {
```
```suggestion
  if (ngDevMode && triggerLView !== null) {
```

AndrewKushnir commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
+    const cleanup = eventManager.onInteraction(element, () => {
+      callback();
+      removeLViewOnDestroy(triggerLView, cleanup);
+      if (initialLView !== triggerLView) {
+        removeLViewOnDestroy(initialLView, cleanup);
+      }
```
👍

AndrewKushnir commented on packages/core/src/render3/instructions/defer_event_manager.ts:
```diff
@@ -0,0 +1,79 @@
+}
+
+/** Manages events related to deferred blocks. */
+export class ɵDeferEventManager {
```
Quick question:  do we plan to reuse this class for other triggers? I think `interaction` and `hover` cases are ok, but for the `viewport` case we'd need an intersection observer, so putting everything in one class would make the code non-tree-shakable.

crisbeto replied:
Yes, I was planning on having two more methods for `hover` and `viewport`, but I actually ended up with having one method. I can also rework it so it's a few top-level function, I just need to verify that we'd be able to inject `NgZone` for `on viewport`.

AndrewKushnir commented on packages/core/src/render3/instructions/defer_event_manager.ts:
```diff
@@ -0,0 +1,79 @@
+      // 3. We're still registering only two events per element, no matter how many deferred blocks
+      // are referencing it.
+      entry = new DeferEventManagerEntry();
+      trigger.addEventListener('click', entry.listener, EVENT_LISTENER_OPTIONS);
+      trigger.addEventListener('keydown', entry.listener, EVENT_LISTENER_OPTIONS);
```
I'd propose moving events to a separate array, so we can run `removeEventListener` using the same list.

crisbeto replied:
I'll give it a try, but my intuition was that it wouldn't be worth the hassle since it's just two events and the `addEventListener` and `removeEventListener` calls are right next to each other so it'll be easy to see that we need to update two places if we were to add another event.

AndrewKushnir reviewed (commented):
@crisbeto looks great! I've left a few comments. I think we'd need to update how we keep track of prefetching triggers before we merge the PR.

crisbeto commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -234,17 +236,36 @@ export function ɵɵdeferPrefetchOnHover() {}  // TODO: implement runtime logic.
+  const tDetails = getTDeferBlockDetails(tView, tNode);
+
+  if (tDetails.loadingState === DeferDependenciesLoadingState.NOT_STARTED) {
+    tDetails.loadingState = DeferDependenciesLoadingState.SCHEDULED;
```
I'm not sure I understand what the issue is here. Aren't both triggers trying to start the preloading of the same resources? Hence we care until one of them fires first after which we can cancel the rest of the triggers. We can move this check inside the `afterRender` to avoid binding any events if `on idle` started it already.

AndrewKushnir replied:
This logic schedules resource loading operation, but doesn't actually initiate the loading process. As a result, there might be multiple prefetch triggers that would result in scheduling multiple operations. Once the first one is triggered, we should cleanup the rest. In some cases (e.g. `on idle` or when we do `on interaction(localRefFromTemplate)`), we do not want to schedule prefetching more than once per defer block type (that may have multiple instances, for ex. if a defer is in a for loop).

I've put together a PR with a proposed change to unify the management of prefetch triggers and cleanup functions: https://github.com/angular/angular/pull/51856. Please let me know what you think about it.

crisbeto commented:
I've addressed all of the feedback, except the one that depends on https://github.com/angular/angular/pull/51856. I also ported over some of the changes from my `on hover` and `on viewport` branch.

crisbeto committed:
feat(core): implement deferred block interaction triggers

Adds the implementation for the `on interaction` and `prefetch on interaction` triggers.

AndrewKushnir reviewed (approved):
👍

dylhunn commented:
This PR was merged into the repository by commit 3cbb2a8ecf202c814c37ab241ce9f57fb574692e.
